actors/evm/src/interpreter/precompiles.rs

```diff
@@ -179,16 +265,150 @@ fn lookup_address<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
 /// Reads a FIL encoded address
 /// Resolves a FIL encoded address into an ID address
 /// returns BE encoded u64 or empty array if nothing found
-fn resolve_address<RT: Runtime>(rt: &RT, input: &[u8]) -> PrecompileResult {
+fn resolve_address<RT: Runtime>(
+    rt: &RT,
+    input: &[u8],
+    o: Option<PrecompileContext>,
+) -> PrecompileResult {
+    assert_eq!(o, None, "no context should reach this precompile");
     let addr = match Address::from_bytes(input) {
         Ok(o) => o,
         Err(_) => return Ok(Vec::new()),
     };
     Ok(rt.resolve_address(&addr).map(|a| a.to_be_bytes().to_vec()).unwrap_or_default())
 }
 
+/// Errors:
+///    TODO should just give 0s?
+/// - `IncorrectInputSize` if offset is larger than total input length
+/// - `InvalidInput` if supplied address bytes isnt a filecoin address
+///
+/// Returns:
+///
+/// `[int256 exit_code, uint codec, uint offset, uint size, []bytes <actor return value>]`
+///
+/// for exit_code:
+/// - negative values are system errors
+/// - positive are user errors (from the called actor)
+/// - 0 is success
+pub fn call_actor<RT: Runtime>(
+    rt: &RT,
+    input: &[u8],
+    ctx: Option<PrecompileContext>,
+) -> PrecompileResult {
+    let ctx = ctx.unwrap();
+
+    // exit_code, codec, offset, size
+    const VALUES_LEN: usize = 4 * 32;
+
+    /// read a u64 from a be encoded "u256" returning an error if u256 bits > 64
+    /// slice must be len 32 or will panic
+    #[inline]
+    fn read_u64(input: &[u8]) -> Result<u64, PrecompileError> {
+        if input[..24] == [0u8; 24] {
+            Err(PrecompileError::InvalidInput)
+        } else {
+            Ok(u64::from_be_bytes(input[24..].try_into().unwrap()))
+        }
+    }
+
+    // TODO Until we support readonly (static) calls at the fvm level, we disallow callactor
+    //      when in static mode as it is sticky and there are no guarantee of preserving the
+    //      static invariant
+    if ctx.is_static {
+        return Err(PrecompileError::CallActorError(StatusCode::StaticModeViolation));
+    }
+
+    // ----- Input Parameters -------
+
+    let input_read = &read_right_pad(input, 32 * 5);
+    let mut input_params: ArrayChunks<u8, 32> = ArrayChunks::new(&input_read);
+
+    let method = read_u64(input_params.next().unwrap())?;
+    let codec = read_u64(input_params.next().unwrap())?;
+    // TODO only CBOR for now
+    if codec != fvm_ipld_encoding::DAG_CBOR {
+        return Err(PrecompileError::InvalidInput);
+    }
+
+    let address_offset = read_u64(input_params.next().unwrap())? as usize;
+    let send_data_offset = read_u64(input_params.next().unwrap())? as usize;
+
+    let address_size = read_u64(input_params.next().unwrap())? as usize;
+    let send_data_size = read_u64(input_params.next().unwrap())? as usize;
```

Comment:
Please read https://docs.soliditylang.org/en/v0.6.11/abi-spec.html#abi.

Reply:
Ok, well after talking we are going to go with just specifying the lengths for now, if solidity is having a particularly hard time decoding these then we can add them defined exactly with solidity a ABI, but this could also be implemented easily enough in solidity itself with an SDK